File: backend/internal/domain/llm.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type LLMProvider string

const (
	ProviderOpenAI    LLMProvider = "openai"
	ProviderOllama    LLMProvider = "ollama"
	ProviderGemini    LLMProvider = "gemini"
	ProviderAnthropic LLMProvider = "anthropic"
)

type LLMConfiguration struct {
	ID        uuid.UUID   `json:"id"`
	Provider  LLMProvider `json:"provider"`
	APIKey    string      `json:"api_key,omitempty"` // Explicitly handle masking in handlers
	BaseURL   string      `json:"base_url,omitempty"`
	Model     string      `json:"model"`
	IsActive  bool        `json:"is_active"`
	CreatedAt time.Time   `json:"created_at"`
	UpdatedAt time.Time   `json:"updated_at"`
}

type LLMClient interface {
	// Generate sends a prompt to the LLM and returns the response.
	// supports streaming via a callback if needed, but for now simple request/response
	Generate(ctx context.Context, prompt string) (string, error)

	// StreamGenerate sends a prompt and streams the response chunks.
	StreamGenerate(ctx context.Context, prompt string, decimals chan<- string) error

	// TestConnection verifies if the client can successfully communicate with the provider.
	TestConnection(ctx context.Context) error

	// ListModels retrieves the list of available models from the provider.
	ListModels(ctx context.Context) ([]string, error)
}

File: frontend/src/api/llm.ts
import { apiClient as client } from './client';

export interface LLMConfiguration {
    id?: string;
    provider: 'openai' | 'ollama' | 'gemini' | 'anthropic';
    api_key: string;
    base_url?: string;
    model: string;
    is_active: boolean;
    created_at?: string;
    updated_at?: string;
}

export const llmApi = {
    getConfig: async (): Promise<LLMConfiguration | null> => {
        const response = await client.get('/settings/llm');
        return response.data;
    },

    updateConfig: async (config: LLMConfiguration): Promise<LLMConfiguration> => {
        const response = await client.put('/settings/llm', config);
        return response.data;
    },

    testConnection: async (config: LLMConfiguration): Promise<{ message: string }> => {
        const response = await client.post('/settings/llm/test', config);
        return response.data;
    },

    listModels: async (config: LLMConfiguration): Promise<string[]> => {
        const response = await client.post('/settings/llm/models', config);
        return response.data.models;
    },

    // Warmup is handled via SSE directly in component usually, or we can use EventSource
    getWarmupEndpoint: () => '/settings/llm/warmup',
};
